lib/galaxy/jobs/runners/hmgm.py

```diff
@@ -0,0 +1,239 @@
+                return
+        else:
+            log.warning("stop_job(): %s: PID %d refuses to die after signaling TERM/KILL" % (job.id, pid))
+    
+    # Run job is a slightly modified version of run_job in runners/local.py.  It builds a command line proc
+    # to execute, reads the stdout and stderr, and returns the status
+    def _run_job(self, job_wrapper):
+        exit_code = 0
+
+        # command line has been added to the wrapper by prepare_job()
+        command_line, exit_code_path = self.__command_line(job_wrapper)
+        job_id = job_wrapper.get_id_tag()
+
+        try:
+            stdout_file = tempfile.NamedTemporaryFile(mode='wb+', suffix='_stdout', dir=job_wrapper.working_directory)
+            stderr_file = tempfile.NamedTemporaryFile(mode='wb+', suffix='_stderr', dir=job_wrapper.working_directory)
+            log.debug('(%s) executing job script: %s' % (job_id, command_line))
+
+            proc = subprocess.Popen(args=command_line,
+                                    shell=True,
+                                    cwd=job_wrapper.working_directory,
+                                    stdout=stdout_file,
+                                    stderr=stderr_file,
+                                    preexec_fn=os.setpgrp)
+
+            proc.terminated_by_shutdown = False
+
+            try:
+                job_wrapper.set_job_destination(job_wrapper.job_destination, proc.pid)
+                job_wrapper.change_state(model.Job.states.RUNNING)
+
+                # Reap the process and get the exit code.
+                exit_code = proc.wait()
+                 
+            except Exception:
+                log.warning("Failed to read exit code from process")
+
+            try:
+                exit_code = int(open(exit_code_path, 'r').read())                    
+            except Exception:
+                log.warning("Failed to read exit code from path %s" % exit_code_path)
+                
+            log.debug("EXIT_CODE_AMP" + str(exit_code))
+            if proc.terminated_by_shutdown:
+                self._fail_job_local(job_wrapper, "job terminated by Galaxy shutdown")
+                return -1
+
+            stdout_file.seek(0)
+            stderr_file.seek(0)
+            self.stdout = shrink_stream_by_size(stdout_file, DATABASE_MAX_STRING_SIZE, join_by="\n..\n", left_larger=True, beginning_on_size_error=True)
+            self.stderr = shrink_stream_by_size(stderr_file, DATABASE_MAX_STRING_SIZE, join_by="\n..\n", left_larger=True, beginning_on_size_error=True)
+            stdout_file.close()
+            stderr_file.close()
+            log.debug('execution finished: %s' % command_line)
+        except Exception:
+            log.exception("failure running job %d", job_wrapper.job_id)
+            self._fail_job_local(job_wrapper, "failure running job")
+            return -1
+        
+        return exit_code
+
```

Comment:
Just a small suggestion: could you mark the change (like begin change, end change) in the "mostly copied" methods, so if we have an upgrade of Galaxy in the future, we can merge our customization more easily?

Reply:
Done